clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -944,23 +941,20 @@ public boolean sameRequest(final OffsetFetchRequestState request) {
         public NetworkClientDelegate.UnsentRequest toUnsentRequest() {
 
             OffsetFetchRequest.Builder builder;
-            if (memberInfo.memberId.isPresent() && memberInfo.memberEpoch.isPresent()) {
-                builder = new OffsetFetchRequest.Builder(
-                        groupId,
-                        memberInfo.memberId.get(),
-                        memberInfo.memberEpoch.get(),
-                        true,
-                        new ArrayList<>(this.requestedPartitions),
-                        throwOnFetchStableOffsetUnsupported);
-            } else {
-                // Building request without passing member ID/epoch to leave the logic to choose
-                // default values when not present on the request builder.
-                builder = new OffsetFetchRequest.Builder(
-                        groupId,
-                        true,
-                        new ArrayList<>(this.requestedPartitions),
-                        throwOnFetchStableOffsetUnsupported);
-            }
+            // Building request without passing member ID/epoch to leave the logic to choose
+            // default values when not present on the request builder.
+            builder = memberInfo.memberEpoch.map(epoch -> new OffsetFetchRequest.Builder(
+                            groupId,
+                            memberInfo.memberId,
+                            epoch,
+                            true,
+                            new ArrayList<>(this.requestedPartitions),
+                            throwOnFetchStableOffsetUnsupported))
+                    .orElseGet(() -> new OffsetFetchRequest.Builder(
+                            groupId,
+                            true,
+                            new ArrayList<>(this.requestedPartitions),
+                            throwOnFetchStableOffsetUnsupported));
```

Comment:
For my understanding, this means that we only set the member id when we have a valid epoch too. In other words, we don't set the member id if we are not actively part of a group. Is my understanding correct?

I wonder if we should do the same for the groupMetadata in order to be consistent. What do you think?

Reply:
I'm not sure about this issue is related to following code or not：
https://github.com/apache/kafka/blob/8c071b02e9908d9facf10c0a18e7e0f9d1b0825f/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L636-L647

If so, I think yes，we could call the method if `memberEpoch` is present.
As my previous comment，the `groupMetadata` should represent the broker acknowledgements since we're moving coordinator logic to broker side.

If I don't misunderstand, I will update it in next commit.

>Hi @lianetm,
>
>Here is my understanding. Please correct me if I’m wrong:
>Since the group management has now moved to the server-side, the ConsumerGroupMetadata basically represents the data acknowledged by the broker. In this case, I think we should keep the ConsumerGroupMetadata.memberId empty at startup. This is because, at that moment, the broker still has no knowledge of the memberId. Once the consumer polls, the broker will then recognize the memberId and update >the assignment, allowing the consumer to update the ConsumerGroupMetadata via MemberStateListener. After >that, ConsumerGroupMetadata will have the memberId.
>
> Does this make sense ?

Reply:
makes sense to me to align and return `memberId` in the consumer group metadata info only if the member is in a group. So after `unsubscribe` (this was my concern btw, not startup), even if internally the `membershipMgr` knows the member ID, we're saying that the consumer group metadata won't include it (change already in place). Agreed.